### ISSUE TITLE
Cluster is blocked because of operation retries in FROZEN state during HotRestart procedure

### ISSUE THREAD

**Issue by yozank** [hazelcast/hazelcast#16021]
(I am not sure if should open this here or in enterprise repo, please let me know so I can move it)

During HotRestart procedure;

- cluster state is set to FROZEN. 
- shutdown one node 
- send calls like `map.size`, `map.get`, etc... 

The cluster becomes unresponsive because the operations are retried due to `TargetNotMemberException, thus block partition operation threads. 

Since we know some partitions do not exist due to the shutdown node, is it possible to reject the operations that target those partitions? 
Or, can we make it optional as some other use-cases may require the operations to be blocked (current behavior) until the dead node rejoins?

**Comment by vojtechtoman** [comment]
@yozank, can you clarify what you mean by "during HotRestart procedure" - do you mean shutting down a member while Hot Restart restore is still in progress on other members? Or is it simply shutting down a member once the restore is complete?

As for blocking the partition threads, note that you can get the very same behavior with 'regular' cluster (i.e. no Hot Restart), as demonstrated by the following test:

```java
@Test
public void test_clusterBlocked_whenFROZEN() {
    HazelcastInstance h1 = factory.newHazelcastInstance();
    HazelcastInstance h2 = factory.newHazelcastInstance();
    warmUpPartitions(h1, h2);
    changeClusterStateEventually(h1, ClusterState.FROZEN);
    h2.shutdown();
    h1.getMap("foo").size();
}
```

**Comment by mmedenjak** [comment]
Just to be clear, the operation retries ("blocking behaviour") is by-design. We retry the operations in hopes that the gone member will return in the meantime.

**Comment by mdogan** [comment]
Due to my understanding what @yozank means is, can we add an option to reject those operations immediately instead of retrying them and blocking the caller.

**Comment by vojtechtoman** [comment]
@mdogan, yes, that is my understanding, too. What I was trying to get at, in a clumsy a way, is whether this potential new option should be specific to Hot Restart or whether it should apply generally (I believe the latter).

**Comment by yozank** [comment]
>do you mean shutting down a member while Hot Restart restore is still in progress on other members?

Yes, this is the condition I mean. 
We can generalize it to FROZEN/PASSIVE state because these states don't allow re-partitioning. 
When a member shuts down, some partitions are lost but the operations to those partitions are blocked because they are retried until the shutdown member is up again. 

As I discussed with Mehmet earlier, and as Matko stated above, this is by design. 
But, a customer has a use-case where they need these calls to be rejected instead of blocking.
So, the question is; is it possible to implement this optional behavior? 

>What I was trying to get at, in a clumsy a way, is whether this potential new option should be specific to Hot Restart or whether it should apply generally (I believe the latter).

Since FROZEN state is used only by HotRestart, I started to explain that way but we can generalize it to the following:

This potential option would be considered when (**the cluster state is FROZEN/PASSIVE (which block re-partitioning)  && there are lost partitions due to member shutdown && there are operations targeting to the lost partitions.**)

**Comment by mmedenjak** [comment]
As mentioned, it's an option we can introduce but currently we're not planning on doing it as we're in the bugfixing phase and the change can be introduced in a minor release. If the severity of the issue is high enough, see with PM if this needs to be fixed in 4.0.
